merlin/models/tf/core/combinators.py

```diff
@@ -514,24 +518,35 @@ def compute_call_output_shape(self, input_shape):
 
     def build(self, input_shape):
         to_prune = []
-        propagate_all_inputs = True
-        if isinstance(input_shape, dict) and all(
-            name in input_shape for name in list(self.parallel_dict.keys())
-        ):
-            propagate_all_inputs = False
 
-        for key, branch in self.parallel_dict.items():
-            branch_shape = input_shape[key] if not propagate_all_inputs else input_shape
-            branch.build(branch_shape)
-            branch_out_shape = branch.compute_output_shape(branch_shape)
-            if self.automatic_pruning and branch_out_shape == {}:
-                to_prune.append(key)
+        for name, layer in self.parallel_dict.items():
+            layer_input_shape = self._maybe_filter_layer_inputs_using_schema(
+                name, layer, input_shape
+            )
+            layer.build(layer_input_shape)
+            layer_out_shape = layer.compute_output_shape(layer_input_shape)
+            if self.automatic_pruning and layer_out_shape == {}:
+                to_prune.append(name)
 
         for p in to_prune:
             self.parallel_layers.pop(p)
 
         return super().build(input_shape)
 
+    def _maybe_filter_layer_inputs_using_schema(self, name, layer, inputs):
+        has_schema = getattr(layer, "has_schema", False)
+        if has_schema and isinstance(inputs, dict):
+            layer_inputs = Filter(layer.schema)(inputs)
```

Comment:
+1 on this. I think the savest would be to extract the logic to filter out the right features of the dict inside this method.

Reply:
Good point. Since we are already checking the existence of schema and whether `inputs` is a dict, the logic becomes very simple, so I opted for explicitly stating the logic in the method.